src/na/device/Device.cpp

```diff
@@ -0,0 +1,465 @@
+  LOCAL_SINGLE_QUBIT,  ///< Local single-qubit operation
+  LOCAL_MULTI_QUBIT,   ///< Local multi-qubit operation
+  SHUTTLING_LOAD,      ///< Shuttling load operation
+  SHUTTLING_MOVE,      ///< Shuttling move operation
+  SHUTTLING_STORE,     ///< Shuttling store operation
+};
+
+/**
+ * @brief Checks if the operation type is a shuttling operation.
+ * @param type The operation type to check.
+ * @return true if the operation type is a shuttling operation, false
+ * otherwise.
+ */
+[[nodiscard]] auto isShuttling(const OperationType type) -> bool {
+  switch (type) {
+  case OperationType::SHUTTLING_LOAD:
+  case OperationType::SHUTTLING_MOVE:
+  case OperationType::SHUTTLING_STORE:
+    return true;
+  default:
+    return false;
+  }
+}
+} // namespace
+
+/**
+ * @brief Implementation of the MQT_NA_QDMI_Device_Session structure.
+ */
+struct MQT_NA_QDMI_Device_Session_impl_d {
+  SessionStatus status = SessionStatus::ALLOCATED;
+};
+
+/**
+ * @brief Implementation of the MQT_NA_QDMI_Device_Job structure.
+ */
+struct MQT_NA_QDMI_Device_Job_impl_d {};
+
+/**
+ * @brief Implementation of the MQT_NA_QDMI_Device_Site structure.
+ */
+struct MQT_NA_QDMI_Site_impl_d {
+  size_t id;        ///< Unique identifier of the site
+  size_t module;    ///< Identifier of the module the site belongs to
+  size_t subModule; ///< Identifier of the sub-module the site belongs to
+  int64_t x;        ///< X coordinate of the site in the lattice
+  int64_t y;        ///< Y coordinate of the site in the lattice
+};
+
+/**
+ * @brief Implementation of the MQT_NA_QDMI_Device_Operation structure.
+ */
+struct MQT_NA_QDMI_Operation_impl_d {
+  std::string name;     ///< Name of the operation
+  OperationType type;   ///< Type of the operation
+  size_t numParameters; ///< Number of parameters for the operation
+  /**
+   * @brief Number of qubits involved in the operation
+   * @note This number is only valid if the operation is a multi-qubit
+   * operation.
+   */
+  size_t numQubits;
+  double duration; ///< Duration of the operation in microseconds
+  double fidelity; ///< Fidelity of the operation
+};
+
+namespace {
+
+/**
+ * @brief Provides access to the device name.
+ * @returns a reference to a static variables that stores the
+ * device name.
+ */
+[[nodiscard]] auto name() -> std::string& {
+  static std::string name;
+  return name;
+}
+
+/**
+ * @brief Provides access to the number of qubits in the device.
+ * @returns a reference to a static variable that stores the number of qubits.
+ */
+[[nodiscard]] auto qubitsNum() -> size_t& {
+  static size_t qubitsNum = 0;
+  return qubitsNum;
+}
+
+/**
+ * @brief Provides access to the lists of sites and operations.
+ * @returns a reference to a static vector of unique pointers to
+ * @ref MQT_NA_QDMI_Site_impl_d.
+ */
+[[nodiscard]] auto sites()
+    -> std::vector<std::unique_ptr<MQT_NA_QDMI_Site_impl_d>>& {
+  static std::vector<std::unique_ptr<MQT_NA_QDMI_Site_impl_d>> sites;
+  return sites;
+}
```

Comment:
At some point we talked about turning the device into a singleton similar to how this could be done for the driver. Did you try that and it was not really worth it, or was there just not enough time? I could still see the potential benefit in doing it.